File: logcat_viewer/parser.py
"""Parse `adb logcat -v threadtime` output lines into structured entries."""
from __future__ import annotations

import re

# Android log priorities (matches android.util.Log constants).
PRIORITY = {"V": 2, "D": 3, "I": 4, "W": 5, "E": 6, "F": 7, "S": 8}
LEVEL_NAMES = {2: "Verbose", 3: "Debug", 4: "Info", 5: "Warn", 6: "Error", 7: "Fatal", 8: "Silent"}

# Unparseable lines are treated as Verbose so the most permissive level still
# shows them, but any raised level threshold hides the noise.
UNKNOWN_PRIORITY = PRIORITY["V"]

# 07-09 14:23:01.123  1234  1250 D BActivityThread: message text
_THREADTIME = re.compile(
    r"^(?P<time>\d{2}-\d{2} \d{2}:\d{2}:\d{2}\.\d+)\s+"
    r"(?P<pid>\d+)\s+(?P<tid>\d+)\s+"
    r"(?P<level>[VDIWEFS])\s+"
    r"(?P<tag>.*?): ?(?P<msg>.*)$"
)


class LogEntry:
    """One parsed logcat line. Uses __slots__ because we hold hundreds of thousands."""

    __slots__ = ("time", "pid", "tid", "level", "priority", "tag", "msg", "raw", "search")

    def __init__(self, time, pid, tid, level, priority, tag, msg, raw):
        self.time = time
        self.pid = pid
        self.tid = tid
        self.level = level
        self.priority = priority
        self.tag = tag
        self.msg = msg
        self.raw = raw
        # Precomputed lowercase haystack for case-insensitive substring filtering.
        self.search = (tag + " " + msg).lower()


def parse_line(line: str) -> LogEntry | None:
    """Return a LogEntry, or None for lines that carry no log content (dividers/blank)."""
    if not line or line.startswith("--------- "):
        return None
    m = _THREADTIME.match(line)
    if m:
        level = m.group("level")
        return LogEntry(
            time=m.group("time"),
            pid=int(m.group("pid")),
            tid=int(m.group("tid")),
            level=level,
            priority=PRIORITY.get(level, UNKNOWN_PRIORITY),
            tag=m.group("tag").rstrip(),
            msg=m.group("msg"),
            raw=line,
        )
    # Non-threadtime line (rare malformed output): keep it, don't lose data.
    return LogEntry(
        time="",
        pid=0,
        tid=0,
        level="?",
        priority=UNKNOWN_PRIORITY,
        tag="",
        msg=line,
        raw=line,
    )

File: androidlab-electron/src/core/parser.ts
/**
 * Parse `adb logcat -v threadtime` output lines into structured entries.
 * Faithful port of logcat_viewer/parser.py.
 */
import { PRIORITY, UNKNOWN_PRIORITY } from './priorities'

/** One parsed logcat line. */
export interface LogEntry {
  time: string
  pid: number
  tid: number
  level: string
  priority: number
  tag: string
  msg: string
  raw: string
  /** Precomputed lowercase haystack for case-insensitive substring filtering. */
  search: string
}

// 07-09 14:23:01.123  1234  1250 D BActivityThread: message text
const THREADTIME =
  /^(?<time>\d{2}-\d{2} \d{2}:\d{2}:\d{2}\.\d+)\s+(?<pid>\d+)\s+(?<tid>\d+)\s+(?<level>[VDIWEFS])\s+(?<tag>[\s\S]*?): ?(?<msg>[\s\S]*)$/

function makeEntry(
  time: string,
  pid: number,
  tid: number,
  level: string,
  priority: number,
  tag: string,
  msg: string,
  raw: string
): LogEntry {
  return {
    time,
    pid,
    tid,
    level,
    priority,
    tag,
    msg,
    raw,
    search: (tag + ' ' + msg).toLowerCase()
  }
}

/** Return a LogEntry, or null for lines that carry no log content (dividers/blank). */
export function parseLine(line: string): LogEntry | null {
  if (!line || line.startsWith('--------- ')) {
    return null
  }
  const m = THREADTIME.exec(line)
  if (m && m.groups) {
    const level = m.groups.level
    return makeEntry(
      m.groups.time,
      parseInt(m.groups.pid, 10),
      parseInt(m.groups.tid, 10),
      level,
      PRIORITY[level] ?? UNKNOWN_PRIORITY,
      m.groups.tag.replace(/\s+$/, ''), // Python .rstrip()
      m.groups.msg,
      line
    )
  }
  // Non-threadtime line (rare malformed output): keep it, don't lose data.
  return makeEntry('', 0, 0, '?', UNKNOWN_PRIORITY, '', line, line)
}
